ngboost/distns/multivariate_normal.py

```diff
@@ -1,252 +1,302 @@
-        mu1_given_0 = mu1 + cov * (1 / (eps + var0)) * (Y - mu0)
-        var0_given_1 = var0 - cov * (1 / (eps + var1)) * cov
-        var1_given_0 = var1 - cov * (1 / (eps + var0)) * cov
-
-        return mu0_given_1, mu1_given_0, var0_given_1, var1_given_0
-
-    def fit(Y):
-        try:
-            # place holder
-            m = np.array([8.0, 8.0])
-            sigma = np.array([[2.0, 1.0], [1.0, 2.0]])
-            L = sp.linalg.cholesky(sigma, lower=True)
-            return np.concatenate([m, L[np.tril_indices(2)]])
-        except:  # NOQA
+            VarComp[:, diag_idx, diag_idx] = value
+            VarComp[:, diag_idx, diag_idx] = value
+            for par_idx in off_diags:
+                q = tril_indices[0][par_idx]
+                k = tril_indices[1][par_idx]
+                if i == k:
+                    value = self.cov[:, q, i] * self.L[:, i, i]
+                    VarComp[:, diag_idx, par_idx] = value
+                    VarComp[:, par_idx, diag_idx] = value
+
+        # Off diagonals  w.r.t. off diagonals
+        for par_idx in off_diags:
+            j = tril_indices[0][par_idx]
+            i = tril_indices[1][par_idx]
+            for par_idx2 in off_diags:
+                k = tril_indices[0][par_idx2]
+                q = tril_indices[1][par_idx2]
+                if i == q:
+                    value = self.cov[:, k, j]
+                    VarComp[:, par_idx, par_idx2] = value
+                    VarComp[:, par_idx2, par_idx] = value
+        return FisherInfo
+
+
+def get_chol_factor(lower_tri_vals):
+    """
+
+    Args:
+        lower_tri_vals: numpy array, shaped as the number of lower triangular
+                        elements, number of observations.
+                        The values ordered according to np.tril_indices(p)
+                        where p is the dimension of the multivariate normal distn
+
+    Returns:
+        Nxpxp numpy array, with the lower triangle filled in. The diagonal is exponentiated.
+
+    """
+    lower_size, N = lower_tri_vals.shape
+
+    # solve p(p+3)/2 = lower_size to get the
+    # number of dimensions.
+
+    p = (-1 + (1 + 8 * lower_size) ** 0.5) / 2
+    p = int(p)
+
+    if not isinstance(lower_tri_vals, np.ndarray):
+        lower_tri_vals = np.array(lower_tri_vals)
+
+    L = np.zeros((N, p, p))
+    for par_ind, (k, l) in enumerate(zip(*np.tril_indices(p))):
+        if k == l:
+            # Add a small number to avoid singular matrices.
+            L[:, k, l] = np.exp(lower_tri_vals[par_ind, :]) + 1e-6
+        else:
+            L[:, k, l] = lower_tri_vals[par_ind, :]
+    return L
+
+
+def MultivariateNormal(k):
+    """
+    #  Factory function that generates classes for
+    #  k-dimensional multivariate normal distributions for NGBoost
+
+    # This distribution has LogScore implemented for it.
+
+    # Currently only for a regression implementation.
+    """
+    assert k >= 2, ValueError(
```

Comment:
I think the thing to do here would be to return `ngboost.distns.Normal`

Reply:
I'll do this shortly. Writing up the natural gradient derivation at the moment too which might make this easier to review. I'll share it tomorrow.

Reply:
I don't think this will be easy, 
I changed this assertion to a warning and returned a Normal which results in:
```
  /home/omalley3/Documents/ProjectOceanVel/ngboost/ngboost/distns/multivariate_normal.py:171: UserWarning: Returning ngboost.distns.Normal as k=1.
    warnings.warn("Returning ngboost.distns.Normal as k=1.")

tests/test_distns.py::test_multivariatenormal[learner0-1]
tests/test_distns.py::test_multivariatenormal[learner1-1]
  /home/omalley3/anaconda3/envs/ngboost/lib/python3.8/site-packages/sklearn/utils/validation.py:72: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
    return f(**kwargs)
``` 

I could make the MVN class be able to handle k=1 (N,1 shaped nd.ndarry for y)  but I think it might be misleading and potentially slower than Normal?

Reply:
Sounds good to me!